include/sparrow/array_data_factory.hpp

```diff
@@ -0,0 +1,347 @@
+        ValueRange&& values,
+        const array_data::bitmap_type& bitmap,
+        std::int64_t offset
+    )
+    {
+        SPARROW_ASSERT_TRUE(values.size() == bitmap.size());
+        SPARROW_ASSERT_TRUE(std::cmp_greater_equal(values.size(), offset));
+
+        const auto create_buffers = [&values]()
+        {
+            using T = std::ranges::range_value_t<ValueRange>;
+            const auto& unwrap_value = [](const T& value)
+            {
+                if constexpr (mpl::is_reference_wrapper_v<T>)
+                {
+                    return value.get();
+                }
+                else
+                {
+                    return value;
+                }
+            };
+
+            std::vector<array_data::buffer_type> buffers(2);
+            buffers[0].resize(sizeof(std::int64_t) * (values.size() + 1), 0);
+            buffers[1].resize(std::accumulate(
+                values.begin(),
+                values.end(),
+                size_t(0),
+                [&unwrap_value](std::size_t res, const auto& s)
+                {
+                    return res + unwrap_value(s).size();
+                }
+            ));
+
+            auto iter = buffers[1].begin();
+            const auto offsets = buffers[0].data<std::int64_t>();
+            size_t i = 0;
+            for (const auto& value : values)
+            {
+                const auto& unwraped_value = unwrap_value(value);
+                SPARROW_ASSERT_TRUE(
+                    std::cmp_less(unwraped_value.size(), std::numeric_limits<std::int64_t>::max())
+                );
+                offsets[i + 1] = offsets[i] + static_cast<std::int64_t>(unwraped_value.size());
+                std::ranges::copy(unwraped_value, iter);
+                std::advance(iter, unwraped_value.size());
+                ++i;
+            }
+            return buffers;
+        };
+
+        using T = std::unwrap_ref_decay_t<std::unwrap_ref_decay_t<std::ranges::range_value_t<ValueRange>>>;
+
+        return {
+            .type = data_descriptor(arrow_type_id<T>()),
+            .length = static_cast<array_data::length_type>(values.size()),
+            .offset = offset,
+            .bitmap = bitmap,
+            .buffers = create_buffers(),
+            .child_data = {},
+            .dictionary = nullptr
+        };
+    }
+
+    template <typename V>
+    struct values_and_indexes
+    {
+        template <mpl::constant_range R>
+            requires std::same_as<std::ranges::range_value_t<R>, std::remove_const_t<V>>
+        explicit values_and_indexes(R&& range)
+        {
+            ranges_to_vec_and_indexes<R>(std::forward<R>(range), *this);
+        }
+
+        std::vector<std::reference_wrapper<V>> values;
+        std::vector<size_t> indexes;
+    };
+
+    template <mpl::constant_range R>
+    void
+    ranges_to_vec_and_indexes(R&& range, values_and_indexes<const std::ranges::range_value_t<R>>& values_and_indexes)
```

Comment:
Do we assume that `values_and_indexes` is empty? If so, state it in the documentation and add an assert (and maybe a `.is_empty()` function in the type?). If not, clarify it and decide if the function should append data or overwrite. Looking at the `reserve` calls, it's not clear what is intended here.

Reply:
Done